File: trunk/zippyzipjp/sample/ruby/getzippyzipjp.rb
#!/usr/bin/ruby
#
# zippyzipjp
# 
# Copyright 2008-2010 Michinobu Maeda.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#      http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require "open-uri"
require "rexml/document"

conf_path = "" + $0
conf_path[/\/[^\/]*$/] = "/getzippyzipjp_conf.rb"
require conf_path

puts Time.now.ctime + " start."

# JSON細切れデータの圧縮ファイルを保存するディレクトリを確認する。
unless (test(?d, ARCDIR))
  unless (0 == Dir.mkdir(ARCDIR))
    puts "Faild to mkdir " + ARCDIR
    exit(false)
  end
end

# JSON細切れデータを保存するディレクトリを確認する。
unless (test(?d, TARGET))
  unless (0 == Dir.mkdir(TARGET))
    puts "Faild to mkdir " + TARGET
    exit(false)
  end
end

# 更新情報を取得する。
text_curr = ""
open(CHECK) {|io|
  text_curr = io.read
}

# 更新されていなければ終了。
PREV = ARCDIR + "/feed.prev"
text_prev = ""
if (test(?e, PREV))
  open(PREV) {|io|
    text_prev = io.read
  }
end

if (text_curr == text_prev)
  puts "exit."
  exit
end

# 更新情報を保存する。
open(PREV, "w") {|io|
  io.write text_curr
}

# 圧縮ファイルを取得して解凍する。
doc = REXML::Document.new text_curr
doc.elements.each("//a") {|element|
  url = element.attributes["href"]
  if (url[/\/json[^\/]+\.zip$/])
    name = "" + url
    name[/.*\//] = ""
    
    zip = ""
    open(url) {|io|
      zip = io.read
    }
    
    path = ARCDIR + "/" + name
    open(path, "w") {|io|
      io.write zip
    }
    
    puts Time.now.ctime + " " + name + " OK " + File.size(path).to_s + " byte"
    system(ZIPCOMMAND + path)
    
  end
}

puts Time.now.ctime + " complete."

File: trunk/zippyzipjp/sample/python/getzippyzipjp.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-
#
# zippyzipjp
# 
# Copyright 2008-2010 Michinobu Maeda.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#      http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os, time, filecmp, shutil, urllib, re, zipfile
from xml.dom.minidom import parse
import const, getzippyzipjp_conf

filenamepttn = re.compile("[^\/]*$")

# JSON細切れデータの圧縮ファイルを保存するディレクトリを確認する。
print time.ctime() + " start."

if not os.path.isdir(const.arcdir):
    os.mkdir(const.arcdir)

if not os.path.isdir(const.arcdir):
    print "Error: Faild to mkdir " + const.arcdir
    os._exit(1)

# JSON細切れデータを保存するディレクトリを確認する。
if not os.path.isdir(const.target):
    os.mkdir(const.target)

if not os.path.isdir(const.target):
    print "Error: Faild to mkdir " + const.arcdir
    os._exit(1)

# 更新情報を取得する。
const.check = const.arcdir + "/" + filenamepttn.search(const.checkUrl, 0).group(0)
const.prev = re.compile(".*\/").search(const.check, 0).group(0) + "prev.txt"
urllib.urlretrieve(const.checkUrl, const.check)

# 更新されていなければ終了。
if os.path.exists(const.prev) and filecmp.cmp(const.check, const.prev):
    print "exit."
    os._exit(0)

# 更新情報を保存する。
shutil.copy(const.check, const.prev)

# 圧縮ファイルを取得して解凍する。
arcpttn = re.compile("\/json[^\/]*\.zip$")
doc = parse(const.prev)

for a in doc.getElementsByTagName("a"):
    
    url = a.getAttribute("href")
    
    if not arcpttn.search(url, 0):
        continue

    name = filenamepttn.search(url, 0).group(0)
    trg = const.arcdir + "/" + name
    urllib.urlretrieve(url, trg)
    
    if not os.path.isfile(trg):
        print "Error: Failed to open archive: " + name
        os._exit(1)
    
    print time.ctime() + " " + name + " OK. " + str(os.path.getsize(trg)) + " byte"
    zip = zipfile.ZipFile(trg, "r")
    
    for ent in zip.namelist():
        f = open(const.target + "/" + ent, 'w')
        f.write(zip.read(ent))
        f.close()
        
    zip.close()

doc.unlink()
print time.ctime() + " complete."
